File: Python3/Coding-Test-Exercise/make_cal.py
from collections import deque


def evaluate(expr: list):
    operation_stack = []
    operation_prior = {'(': 0, ')': 0, '*': 1, '+': 2, '-': 2}
    postfix = deque()
    for item in expr:
        if item.isdecimal():
            postfix.append(item)
        elif item == '(':
            operation_stack.append(item)
        elif item == ')':
            while (top := operation_stack.pop()) != '(':
                postfix.append(top)
        else:
            while len(operation_stack) != 0:
                top = operation_stack[-1]
                if top != '(' and operation_prior[top] <= operation_prior[item]:
                    operation_stack.pop()
                    postfix.append(top)
                else:
                    break
            operation_stack.append(item)

    while len(operation_stack) != 0:
        item = operation_stack.pop()
        postfix.append(item)

    eval_stack = []
    while len(postfix) != 0:
        value = postfix.popleft()
        if value.isdecimal():
            eval_stack.append(int(value))
        else:
            right = eval_stack.pop()
            left = eval_stack.pop()

            if value == '+':
                eval_stack.append(left + right)
            elif value == '-':
                eval_stack.append(left - right)
            else:
                eval_stack.append(left * right)

    return eval_stack[0]


test_case = int(input())
for _ in range(test_case):
    _ = int(input())
    expressions = list(input().strip().split(" "))
    print(evaluate(expressions))

File: C++/Coding_Test_Exercise/make_cal.cpp
// Why Rumtime Error?
#include <iostream>
#include <string>
#include <stack>
#include <queue>
using namespace std;

int getPrior(string);
int eval(string*, int);

int main()
{
    int testCases;
    cin >> testCases;
    for (int i = 0; i < testCases; i++)
    {
        int length;
        cin >> length;
        cin.ignore();
        string expression;
        getline(cin, expression);
        string exprArray[length];
        int arrayPtr = 0;
        string item;
        for (int j = 0; j < expression.length(); j++)
        {
            if (expression[j] != ' ')
            {
                item += expression[j];
                if (j == expression.length() - 1)
                {
                    exprArray[arrayPtr++] = item;
                    item = "";
                }
            }
            else
            {
                exprArray[arrayPtr++] = item;
                item = "";
            }
        }
        cout << eval(exprArray, length) << endl;
    }
    return 0;
}

int getPrior(string item)
{
    if (item == "+" || item == "-") { return 2; }
    else if (item == "*") { return 1; }
    else { return 0; }
}

int eval(string* expr, int size)
{
    stack<string> operationStack;
    queue<string> postfix;
    for (int i = 0; i < size; i++)
    {
        if (atoi(expr[i].c_str()) != 0 || expr[i].compare("0") == 0) { postfix.push(expr[i]); }
        else if (expr[i] == "(") { operationStack.push(expr[i]); }
        else if (expr[i] == ")")
        {
            while (operationStack.top() != "(")
            {
                string operat = operationStack.top();
                operationStack.pop();
                postfix.push(operat);
            }
            operationStack.pop();
        }
        else
        {
            if (operationStack.empty() || operationStack.top() == "(") { operationStack.push(expr[i]); }
            else
            {
                while (!operationStack.empty())
                {
                    string operat = operationStack.top();
                    if (operat != "(" && getPrior(operat) <= getPrior(expr[i]))
                    {
                        operationStack.pop();
                        postfix.push(operat);
                    }
                    else { break; }
                }
                operationStack.push(expr[i]);
            }
        }
    }

    while (!operationStack.empty())
    {
        string operat = operationStack.top();
        operationStack.pop();
        postfix.push(operat);
    }

    stack<int> evalStack;
    while (!postfix.empty())
    {
        string value = postfix.front();
        postfix.pop();

        if (atoi(value.c_str()) != 0 || value.compare("0") == 0) { evalStack.push(atoi(value.c_str())); }
        else
        {
            int right = evalStack.top();
            evalStack.pop();
            int left = evalStack.top();
            evalStack.pop();

            if (value == "+") { evalStack.push(left + right); }
            else if (value == "-") { evalStack.push(left - right); }
            else { evalStack.push(left * right); }
        }
    }
    return evalStack.top();
}